File: go-cli/internal/fsutil/fsutil.go
// Package fsutil ports cli/src/fsutil.ts: template-dir copy (with Handlebars
// rendering of .hbs files via raymond), the anchor-comment insert mechanism,
// and the JSON / TOML-array merge helpers. Behavior is a deliberate 1:1 port,
// including the narrow regex-based TOML edit and the "skip undefined values"
// JSON-merge rule (see mergeManifestDeps in compose.go), so both CLIs treat
// the same registry the same way.
package fsutil

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/aymerick/raymond"
)

func PathExists(p string) bool {
	_, err := os.Stat(p)
	return err == nil
}

func IsDir(p string) bool {
	info, err := os.Stat(p)
	return err == nil && info.IsDir()
}

// CopyTemplateDir recursively copies src into dest. Files ending in .hbs are
// rendered with raymond (Handlebars) using vars and written without the .hbs
// suffix; everything else is copied byte-for-byte. template.config.json /
// page.config.json are never copied (they're registry metadata, not project
// files).
func CopyTemplateDir(src, dest string, vars map[string]interface{}) error {
	entries, err := os.ReadDir(src)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(dest, 0755); err != nil {
		return err
	}
	for _, entry := range entries {
		name := entry.Name()
		if name == "template.config.json" || name == "page.config.json" {
			continue
		}
		from := filepath.Join(src, name)
		to := filepath.Join(dest, name)
		if entry.IsDir() {
			if err := CopyTemplateDir(from, to, vars); err != nil {
				return err
			}
			continue
		}
		if strings.HasSuffix(name, ".hbs") {
			raw, err := os.ReadFile(from)
			if err != nil {
				return err
			}
			rendered, err := raymond.Render(string(raw), vars)
			if err != nil {
				return fmt.Errorf("rendering %s: %w", from, err)
			}
			if err := os.WriteFile(to[:len(to)-4], []byte(rendered), 0644); err != nil {
				return err
			}
			continue
		}
		if err := copyFile(from, to); err != nil {
			return err
		}
	}
	return nil
}

func copyFile(from, to string) error {
	info, err := os.Stat(from)
	if err != nil {
		return err
	}
	data, err := os.ReadFile(from)
	if err != nil {
		return err
	}
	return os.WriteFile(to, data, info.Mode().Perm())
}

// InsertAtAnchor inserts content immediately above the line containing anchor,
// indenting it with whatever leading text preceded the anchor on that line,
// and leaves the anchor line itself untouched so later composes can insert
// again at the same spot. content may itself contain embedded "\n" for a
// multi-line insert; only its first line receives the indent (matches the TS
// implementation, which splices one array element that happens to contain
// newlines and then joins the whole file back with "\n").
func InsertAtAnchor(filePath, anchor, content string) error {
	raw, err := os.ReadFile(filePath)
	if err != nil {
		return err
	}
	lines := strings.Split(string(raw), "\n")
	idx := -1
	for i, l := range lines {
		if strings.Contains(l, anchor) {
			idx = i
			break
		}
	}
	if idx == -1 {
		return fmt.Errorf("anchor %q not found in %s", anchor, filePath)
	}
	indent := lines[idx][:strings.Index(lines[idx], anchor)]
	inserted := indent + content
	lines = append(lines, "")
	copy(lines[idx+1:], lines[idx:])
	lines[idx] = inserted
	return os.WriteFile(filePath, []byte(strings.Join(lines, "\n")), 0644)
}

// InsertAtAnchorSafe is the best-effort variant used for optional wiring
// targets (e.g. a nav component that only exists if some other feature/page
// installed it). Returns whether the insert happened instead of erroring.
func InsertAtAnchorSafe(filePath, anchor, content string) bool {
	return InsertAtAnchor(filePath, anchor, content) == nil
}

// writeJSON writes v as 2-space-indented JSON with a trailing newline and no
// HTML-escaping (package.json scripts routinely contain "&&", which Go's
// default encoder would otherwise mangle into &).
func writeJSON(filePath string, v interface{}) error {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	enc.SetIndent("", "  ")
	if err := enc.Encode(v); err != nil {
		return err
	}
	return os.WriteFile(filePath, buf.Bytes(), 0644)
}

func WriteJSONFile(filePath string, v interface{}) error {
	return writeJSON(filePath, v)
}

// MergeJsonFile ports mergeJsonFile from fsutil.ts: for each key in patch,
// nil/absent values are skipped entirely (this is the exact bug fix being
// replicated — merging an absent value must never clobber a key already
// merged by an earlier feature), object values are shallow-merged into the
// existing object at that key, everything else overwrites.
func MergeJsonFile(filePath string, patch map[string]interface{}) error {
	raw, err := os.ReadFile(filePath)
	if err != nil {
		return err
	}
	var doc map[string]interface{}
	if err := json.Unmarshal(raw, &doc); err != nil {
		return err
	}
	for key, value := range patch {
		if value == nil {
			continue
		}
		if obj, ok := asStringMap(value); ok {
			existing, _ := asStringMap(doc[key])
			if existing == nil {
				existing = map[string]interface{}{}
			}
			for k, v := range obj {
				existing[k] = v
			}
			doc[key] = existing
		} else {
			doc[key] = value
		}
	}
	return writeJSON(filePath, doc)
}

func asStringMap(v interface{}) (map[string]interface{}, bool) {
	switch m := v.(type) {
	case map[string]interface{}:
		return m, true
	case map[string]string:
		out := make(map[string]interface{}, len(m))
		for k, vv := range m {
			out[k] = vv
		}
		return out, true
	default:
		return nil, false
	}
}

var pkgNameSplit = regexp.MustCompile(`[=<>~!\[; ]`)

func pkgNameOf(spec string) string {
	parts := pkgNameSplit.Split(spec, 2)
	return strings.ToLower(strings.TrimSpace(parts[0]))
}

// MergeTomlDependencyArray ports mergeTomlDependencyArray from fsutil.ts: a
// deliberately narrow regex edit (not a full TOML parse) of a single
// contiguous `key = [ ... ]` array, skipping items already present by
// package name (ignoring version specifiers).
func MergeTomlDependencyArray(filePath, key string, items []string) error {
	if len(items) == 0 {
		return nil
	}
	raw, err := os.ReadFile(filePath)
	if err != nil {
		return err
	}
	text := string(raw)
	re := regexp.MustCompile(regexp.QuoteMeta(key) + `\s*=\s*\[([^\]]*)\]`)
	loc := re.FindStringSubmatchIndex(text)
	if loc == nil {
		return fmt.Errorf("could not find %q = [...] in %s", key, filePath)
	}
	matchStart, matchEnd := loc[0], loc[1]
	groupStart, groupEnd := loc[2], loc[3]
	group := text[groupStart:groupEnd]

	strRe := regexp.MustCompile(`"([^"]+)"`)
	existingMatches := strRe.FindAllStringSubmatch(group, -1)
	existing := make([]string, 0, len(existingMatches))
	existingNames := map[string]bool{}
	for _, m := range existingMatches {
		existing = append(existing, m[1])
		existingNames[pkgNameOf(m[1])] = true
	}

	var toAdd []string
	for _, item := range items {
		if !existingNames[pkgNameOf(item)] {
			toAdd = append(toAdd, item)
		}
	}
	if len(toAdd) == 0 {
		return nil
	}

	merged := append(append([]string{}, existing...), toAdd...)
	var sb strings.Builder
	sb.WriteString(key)
	sb.WriteString(" = [\n")
	for _, m := range merged {
		sb.WriteString("    \"")
		sb.WriteString(m)
		sb.WriteString("\",\n")
	}
	sb.WriteString("]")

	rebuilt := text[:matchStart] + sb.String() + text[matchEnd:]
	return os.WriteFile(filePath, []byte(rebuilt), 0644)
}

type EnvVarLike struct {
	Key     string
	Value   string
	Comment string
}

// AppendEnvVars ports appendEnvVars from fsutil.ts.
func AppendEnvVars(filePath string, vars []EnvVarLike) error {
	if len(vars) == 0 {
		return nil
	}
	var existing string
	if PathExists(filePath) {
		raw, err := os.ReadFile(filePath)
		if err != nil {
			return err
		}
		existing = string(raw)
	}
	lines := make([]string, 0, len(vars))
	for _, v := range vars {
		if v.Comment != "" {
			lines = append(lines, fmt.Sprintf("# %s\n%s=%s", v.Comment, v.Key, v.Value))
		} else {
			lines = append(lines, fmt.Sprintf("%s=%s", v.Key, v.Value))
		}
	}
	sep := ""
	if existing != "" && !strings.HasSuffix(existing, "\n") {
		sep = "\n"
	}
	return os.WriteFile(filePath, []byte(existing+sep+strings.Join(lines, "\n")+"\n"), 0644)
}

File: cli/src/fsutil.ts
import fs from "node:fs/promises";
import path from "node:path";
import Handlebars from "handlebars";

/** Recursively copy `src` into `dest`. Files ending in `.hbs` are compiled with
 * `vars` and written without the `.hbs` suffix; everything else is copied verbatim. */
export async function copyTemplateDir(
  src: string,
  dest: string,
  vars: Record<string, string>,
): Promise<void> {
  const entries = await fs.readdir(src, { withFileTypes: true });
  await fs.mkdir(dest, { recursive: true });
  for (const entry of entries) {
    if (entry.name === "template.config.json" || entry.name === "page.config.json") continue;
    const from = path.join(src, entry.name);
    const to = path.join(dest, entry.name);
    if (entry.isDirectory()) {
      await copyTemplateDir(from, to, vars);
    } else if (entry.name.endsWith(".hbs")) {
      const raw = await fs.readFile(from, "utf-8");
      const rendered = Handlebars.compile(raw)(vars);
      await fs.writeFile(to.slice(0, -4), rendered);
    } else {
      await fs.copyFile(from, to);
    }
  }
}

export async function pathExists(p: string): Promise<boolean> {
  return fs
    .stat(p)
    .then(() => true)
    .catch(() => false);
}

/** Insert `content` immediately above the line containing `anchor`, keeping the
 * anchor in place so later composes can insert again at the same spot. */
export async function insertAtAnchor(
  filePath: string,
  anchor: string,
  content: string,
): Promise<void> {
  const text = await fs.readFile(filePath, "utf-8");
  const lines = text.split("\n");
  const idx = lines.findIndex((l) => l.includes(anchor));
  if (idx === -1) {
    throw new Error(`Anchor "${anchor}" not found in ${filePath}`);
  }
  const indent = lines[idx].slice(0, lines[idx].indexOf(anchor));
  lines.splice(idx, 0, `${indent}${content}`);
  await fs.writeFile(filePath, lines.join("\n"));
}

/** Best-effort variant of insertAtAnchor for optional wiring targets (e.g. a nav
 * component that only exists if the dashboard page was installed). Returns
 * whether the insert happened instead of throwing. */
export async function insertAtAnchorSafe(
  filePath: string,
  anchor: string,
  content: string,
): Promise<boolean> {
  try {
    await insertAtAnchor(filePath, anchor, content);
    return true;
  } catch {
    return false;
  }
}

export async function mergeJsonFile(
  filePath: string,
  patch: Record<string, unknown>,
): Promise<void> {
  const raw = await fs.readFile(filePath, "utf-8");
  const json = JSON.parse(raw);
  for (const [key, value] of Object.entries(patch)) {
    if (value === undefined) continue;
    if (value && typeof value === "object" && !Array.isArray(value)) {
      json[key] = { ...(json[key] ?? {}), ...value };
    } else {
      json[key] = value;
    }
  }
  await fs.writeFile(filePath, JSON.stringify(json, null, 2) + "\n");
}

const pkgNameOf = (spec: string) => spec.split(/[=<>~!\[; ]/)[0].trim().toLowerCase();

/** Appends pip-style requirement strings into a pyproject.toml `dependencies = [...]`
 * array, skipping packages already listed (matched on name, ignoring version specs).
 * A deliberately narrow regex edit rather than a full TOML parse — safe as long as
 * the array is a single contiguous `key = [ ... ]` block, which every base template
 * here controls. */
export async function mergeTomlDependencyArray(filePath: string, key: string, items: string[]): Promise<void> {
  if (items.length === 0) return;
  const text = await fs.readFile(filePath, "utf-8");
  const re = new RegExp(`${key}\\s*=\\s*\\[([^\\]]*)\\]`, "s");
  const match = text.match(re);
  if (!match) throw new Error(`Could not find "${key} = [...]" in ${filePath}`);

  const existing = Array.from(match[1].matchAll(/"([^"]+)"/g)).map((m) => m[1]);
  const existingNames = new Set(existing.map(pkgNameOf));
  const toAdd = items.filter((i) => !existingNames.has(pkgNameOf(i)));
  if (toAdd.length === 0) return;

  const merged = [...existing, ...toAdd];
  const rebuilt = `${key} = [\n${merged.map((m) => `    "${m}",`).join("\n")}\n]`;
  await fs.writeFile(filePath, text.slice(0, match.index) + rebuilt + text.slice((match.index ?? 0) + match[0].length));
}

export async function appendEnvVars(
  filePath: string,
  vars: { key: string; value: string; comment?: string }[],
): Promise<void> {
  if (vars.length === 0) return;
  const existing = (await pathExists(filePath)) ? await fs.readFile(filePath, "utf-8") : "";
  const lines = vars.map((v) => (v.comment ? `# ${v.comment}\n${v.key}=${v.value}` : `${v.key}=${v.value}`));
  const sep = existing && !existing.endsWith("\n") ? "\n" : "";
  await fs.writeFile(filePath, existing + sep + lines.join("\n") + "\n");
}

const PM_INSTALL: Record<string, [string, string[]]> = {
  npm: ["npm", ["install"]],
  pnpm: ["pnpm", ["install"]],
  yarn: ["yarn", []],
  bun: ["bun", ["install"]],
};

/** Maps a package manager name to its install command + args. Falls back to npm
 * for anything unrecognized rather than throwing — scaffolding shouldn't hard-fail
 * over a typo'd --pm value. */
export function installCommandFor(packageManager: string | undefined): [string, string[]] {
  return PM_INSTALL[packageManager ?? "npm"] ?? PM_INSTALL.npm;
}
